src/wasm/wasm-binary.cpp

```diff
@@ -24,15 +24,40 @@
 namespace wasm {
 
 void WasmBinaryWriter::prepare() {
-  // we need function types for all our functions
-  for (auto& func : wasm->functions) {
-    if (func->type.isNull()) {
-      func->type = ensureFunctionType(getSig(func.get()), wasm)->name;
+  // Traverse binary to collect types and their frequency
+  struct TypeCounter : WalkerPass<PostWalker<TypeCounter>> {
```

Comment:
This will traverse the entire module using a single core, please parallelize it (the one tricky thing is that the map keys must not be modified in parallel).

Reply:
Hmm, maybe I'm missing something, but I think we can do things similar to other passes for safe parallelism - like store the data in a function-local info object, then merge after the parallel execution. But this lgtm for landing, I can do that improvement as a refactoring, I have an idea for a helper class.